# STAT 207 Lab Assignment 2 - Individual Part - [25 total points] - Solutions

## Dataframe Manipulation and Cleaning

Due: Tuesday, September 5, 11:59pm CST

<hr>


## <u>Case Study</u>: Steam Game Play and Purchase Analysis
Steam is the world's most popular PC Gaming hub. With a massive collection that includes everything from AAA blockbusters to small indie titles, great discovery tools can be super valuable for Steam. We would like to use the following dataset below to learn more about Steam player behavior.

The dataset we will be analyzing is comprised of a random sample of 500 steam users and their game play and purchase behaviors. It has the following columns:
* **user_id**
* **game_name**,
* **activity**:
    - purchase: indicating that the user has *purchased* the corresponding game
    - play: indicating that the user has *played* the corresponding game (for at least some amount of time.) 
* **hours_played_if_play**:
    - if the row corresponds to a 'play' activity, this number represents the number of hours the user has played the game
    - if the row corresponds to a 'purchase' activity, this number is always a 1 (and means nothing... it's a placeholder).

<hr>


### <u>Research Goals</u>

In this analysis, we will consider a game's **"all-star" player** to be the person who had played this game for the most amount of hours.

1. **Missing Value Cleaning**: Identify any implicit or explicit missing values in the dataset and drop any rows that have a missing value.
2. **Game Purchase vs. Play Behaviors**: What percent of games that were *bought* were actually *played* in this dataset?
3. **Top Game Buyer**: Learn more about the person who bought the most amount of games in this dataset. How many of these games did they actually play?
4. **"All Star Players"**: Who were the "all-star players" of the games? How long did they play this game?
5. **All Star Player Behavior**: What percent of games did that these "all star players" bought did they actually play?



### <u>Points </u>
<p style="margin-top:0in;margin-right:0in;margin-bottom:0in;margin-left:0in;font-size:11.0pt;font-family:&quot;Calibri&quot;,sans-serif;line-height:normal;"></p>
<table style="border: none;border-collapse: collapse;width:102pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:51pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">6</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">9</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
    </tbody>
</table>
<p></p>




<hr>

### <u>STAT107 Review</u>
Part of this individual assignment will involve a review of STAT107 code that you have seen before including the following. Click on the links below if you need a refresher.

1. <a href="https://discovery.cs.illinois.edu/learn/Basics-of-Data-Science-with-Python/DataFrames-with-Conditionals/">Row Filtering</a>
2. <a href="https://discovery.cs.illinois.edu/guides/Modifying-DataFrames/dataframe-groupby/">The groupby() and reset_index() Functions</a>
2. <a href="https://discovery.cs.illinois.edu/guides/Modifying-DataFrames/sorting-a-dataframe-with-pandas/">The sort_values() Function</a>


### <u>Tutorial you Will Need</u>: Utilizing the *.shape* attribute

The **.shape** attribute for a given dataframe returns a **tuple** of two values, the first value in the tuple is the number of rows, the second value is the number of columns.

In [1]:
import pandas as pd
demo_df=pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
demo_df

a  b
0  1  4
1  2  5
2  3  6

We can index a tuple with 0 and 1 indices to extract the number of rows and columns from the dataframe separately.

In [2]:
#Number of rows and columns
demo_df.shape

(3, 2)

In [3]:
#Number of rows
demo_df.shape[0]

3

In [4]:
#Number of columns
demo_df.shape[1]

2

<hr>

### Imports

In [5]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### 1. Read the csv
First, read the **steam_sample_500.csv** file into a dataframe and display the first five rows of the dataframe.

In [6]:
steam_smp500 = pd.read_csv('steam_sample_500.csv', na_values='data suppressed')

### 2. How many rows does the dataframe have? (Use Python code to show this).

In [7]:
steam_smp500.shape

(8601, 4)

### 3. Data cleaning and data type checking

We plan to analyze the **"hours_played_if_play"** variable in this assignment. In order to ensure that our subsequent functions work appropriately on this numerical variable, you should do the following.
1. Identify any values (if any) that have been encoded in the csv to represent a **missing value**. 
2. Make sure that the values in your **"hours_played_if_play"** variable are represented as numerical values (either float64 or int64 object types). If not, fix this using the methods we discussed in class.
3. Delete any rows in your dataframe that have missing values (both explicit or implict). Remember, some "missing values", in the way that they're currently encoded in the csv may not be initially detectable by Python. It is your job to figure out what these implicit missing values are and make it so Python does detect them. And then delete any rows that have them.

In [8]:
steam_smp500.head()

user_id    game_name  activity hours_played_if_play
0  218323237       Dota 2  purchase                    1
1  218323237       Dota 2      play                 14.8
2  218323237  Tomb Raider  purchase                    1
3  218323237  Tomb Raider      play                  8.8
4  218323237         TERA  purchase                    1

In [9]:
steam_smp500.dtypes

user_id                  int64
game_name               object
activity                object
hours_played_if_play    object
dtype: object

In [10]:
steam_smp500.isna()

user_id  game_name  activity  hours_played_if_play
0       False      False     False                 False
1       False      False     False                 False
2       False      False     False                 False
3       False      False     False                 False
4       False      False     False                 False
...       ...        ...       ...                   ...
8596    False      False     False                 False
8597    False      False     False                 False
8598    False      False     False                 False
8599    False      False     False                 False
8600    False      False     False                 False

[8601 rows x 4 columns]

In [11]:
steam_smp500.isna().sum()

user_id                 0
game_name               0
activity                0
hours_played_if_play    8
dtype: int64

In [12]:
steam_smp500.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8601 entries, 0 to 8600
Data columns (total 4 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   user_id               8601 non-null   int64 
 1   game_name             8601 non-null   object
 2   activity              8601 non-null   object
 3   hours_played_if_play  8593 non-null   object
dtypes: int64(1), object(3)
memory usage: 268.9+ KB


In [13]:
unique_values = steam_smp500['hours_played_if_play'].unique()
unique_values


array(['1', '14.8', '8.8', '0.9', '1784', '626', '176', '69', '48',
       '16.1', '13.6', '11.7', '10.2', '5.8', '4.1', '2.3', '2.2', '0.6',
       '0.4', '0.3', '1118', '17.2', '5.9', '22', '86', nan, '1.5', '1.2',
       '0.8', '290', '60', '14.1', '12.5', '9', '7.3', '4.8', '3.2', '3',
       '0.7', '0.5', '2.1', '191', '154', '100', '65', '64', '12.8',
       '9.7', '3.5', '2.6', '6.9', '1395', '25', '12.3', '10.6', '8.2',
       '6.3', '6', '5.4', '4', '1.8', '0.1', '110', '299', '21', '1.9',
       '258', '239', '45', '1.1', '10.7', '2.7', '30', '443', '23', '9.1',
       '2526', '1906', '725', '0.2', '2.4', '608', '337', '115', '95',
       '90', '88', '87', '74', '72', '70', '63', '58', '53', '52', '51',
       '47', '43', '37', '35', '34', '31', '29', '28', '27', '26', '24',
       '20', '19.6', '19', '17.8', '17', '15.8', '15', '13.8', '13.2',
       '12.1', '12', '10.5', '9.8', '9.4', '9.3', '8.4', '7.4', '7',
       '6.8', '6.4', '5.7', '5.1', '5', '4.6', '4.3', '3.7', '3.

In [14]:
steam_smp500['hours_played_if_play'] = pd.to_numeric(steam_smp500['hours_played_if_play'], errors='coerce')

In [15]:
SS500_cleaned = steam_smp500.dropna(subset=['hours_played_if_play'])
SS500_cleaned


user_id                                     game_name  activity  \
0     218323237                                        Dota 2  purchase   
1     218323237                                        Dota 2      play   
2     218323237                                   Tomb Raider  purchase   
3     218323237                                   Tomb Raider      play   
4     218323237                                          TERA  purchase   
...         ...                                           ...       ...   
8596    7519923  Counter-Strike Condition Zero Deleted Scenes  purchase   
8597  103963922                               Team Fortress 2  purchase   
8598  103963922                               Team Fortress 2      play   
8599  300991661                                        Dota 2  purchase   
8600  300991661                                        Dota 2      play   

      hours_played_if_play  
0                      1.0  
1                     14.8  
2                      1.0  
3                      8.8  
4                      1.0  
...                    ...  
8596                   1.0  
8597                   1.0  
8598                   0.4  
8599                   1.0  
8600                   0.3  

[8591 rows x 4 columns]

In [16]:
SS500_cleaned[~((SS500_cleaned['hours_played_if_play'] == 1) & (SS500_cleaned['activity'] == 'purchase'))]


user_id                        game_name activity  \
1     218323237                           Dota 2     play   
3     218323237                      Tomb Raider     play   
5     218323237                             TERA     play   
7     126340495                           Dota 2     play   
9     126340495  Counter-Strike Global Offensive     play   
...         ...                              ...      ...   
8590  138590829                           Dota 2     play   
8592  190460990                           Dota 2     play   
8594    7519923                   Counter-Strike     play   
8598  103963922                  Team Fortress 2     play   
8600  300991661                           Dota 2     play   

      hours_played_if_play  
1                     14.8  
3                      8.8  
5                      0.9  
7                   1784.0  
9                    626.0  
...                    ...  
8590                1026.0  
8592                   1.5  
8594                  35.0  
8598                   0.4  
8600                   0.3  

[3197 rows x 4 columns]

In [17]:
SS500_cleaned.info()
SS500_cleaned.head()


<class 'pandas.core.frame.DataFrame'>
Index: 8591 entries, 0 to 8600
Data columns (total 4 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   user_id               8591 non-null   int64  
 1   game_name             8591 non-null   object 
 2   activity              8591 non-null   object 
 3   hours_played_if_play  8591 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 335.6+ KB


user_id    game_name  activity  hours_played_if_play
0  218323237       Dota 2  purchase                   1.0
1  218323237       Dota 2      play                  14.8
2  218323237  Tomb Raider  purchase                   1.0
3  218323237  Tomb Raider      play                   8.8
4  218323237         TERA  purchase                   1.0

### 4. Number of Dropped Rows 

How many rows with missing values did you delete from your original dataframe?

In [18]:
dropped = len(steam_smp500) - len(SS500_cleaned)

In [19]:
dropped

10

### 5. Separate the dataframes. 

Next, in this analysis we would like to analyze the game purchase information and game play information separately. First create two new dataframes:
* one that is comprised of the purchase rows
* the other that is comprised of the play rows.

In [20]:
purchase = SS500_cleaned[SS500_cleaned.activity == 'purchase']
#play = SS500_cleaned[~(SS500_cleaned.activity == 'purchase')]
play = SS500_cleaned[SS500_cleaned.activity == 'play']
purchase.shape, play.shape

((5394, 4), (3197, 4))

In [21]:
purchase.head(), play.head()


(     user_id                        game_name  activity  hours_played_if_play
 0  218323237                           Dota 2  purchase                   1.0
 2  218323237                      Tomb Raider  purchase                   1.0
 4  218323237                             TERA  purchase                   1.0
 6  126340495                           Dota 2  purchase                   1.0
 8  126340495  Counter-Strike Global Offensive  purchase                   1.0,
      user_id                        game_name activity  hours_played_if_play
 1  218323237                           Dota 2     play                  14.8
 3  218323237                      Tomb Raider     play                   8.8
 5  218323237                             TERA     play                   0.9
 7  126340495                           Dota 2     play                1784.0
 9  126340495  Counter-Strike Global Offensive     play                 626.0)

### 6. Percentage of Game Purchases that are Actually Played

What percentage of all game purchases in this dataset are actually played by the player that purchased them?

*You can assume that a game and user id that has a 'play' row ALSO MUST have a 'purchase' row (but not the other way around).*

**For full credit, use automated Python code to find your answer. (Ie. don't put something like '30/50', where you found the 30 and 50 from the output of other code.)**

In [22]:
percent_played = (len(play)/len(purchase))*100
percent_played


59.26955876900259

### 7. Top Game Purchaser

Next, we'd like to learn more about the people who *purchase* the most amount of games.

#### 7.1. Most Amount of Games Purchased

What is the user id of the player who has purchased the most amount of games in this dataset, and how many games have they purchased?

In [23]:
top_purchase = purchase['user_id'].value_counts().head(1)
top_purchase_user_id = top_purchase.index[0]
top_purchase_count = top_purchase.values[0]

In [24]:
top_purchase_user_id, top_purchase_count

(12610800, 292)

The user id is 12610800; This user has purchased 292 games.

#### 7.2. How many games has this user actually played?

What percentage of the purchased games did this player from 7.1 actually play?

In [25]:
top_user_play_rate = ((play[play.user_id == top_purchase_user_id].shape[0])/top_purchase_count)*100
top_user_play_rate

56.5068493150685

## 8. Examining Some Game "All-Star" Players

Next, we'd like to examine the play vs. purchase behavior of a few games' "all star players". We will consider a game's "all star player" to be the one who has played the game the longest in this dataset.

### 8.1. Each Game's Highest Play Time

In a single dataframe below, display the highest amount of time that each game was played.

For full credit, your final resulting dataframe should:
1. Be comprised of *two columns*
 * game_name
 * hours_played_if_play
2. And should have an index going from 0,1,2,... etc.

In [26]:
AS_players = play.groupby('game_name')['hours_played_if_play'].max().reset_index()
AS_players.columns = ['game_name', 'hours_played_if_play']

In [27]:
AS_players.head()

game_name  hours_played_if_play
0      100% Orange Juice                   0.9
1          7 Days to Die                 336.0
2                8BitBoy                  24.0
3                8BitMMO                   0.2
4  A Valley Without Wind                   1.2

### 8.2. Ranking the Games by Highest Maximum Game Play Time

Sort your final resulting dataframe from 8.1 by the number of hours that the game was played by the all-star player *in descending order*.

Overwrite your dataframe, so it remains sorted in this order.

In [28]:
AS_players = AS_players.sort_values(by='hours_played_if_play', ascending=False).reset_index(drop=True)

### 8.3. Top 2 Games with the Most Dedicated "All-Star Player"

Which two games in this dataset had the most "dedicated" "all star players"? Or in other words, which two games were played the longest amount of time by a particular player?

In [29]:
top_2 = AS_players.head(2)
top_2['game_name']
top_2

game_name  hours_played_if_play
0                      Dota 2                5608.0
1  Sid Meier's Civilization V                5002.0

From what I can gather, it seems that the top two games from our "All-Star Players" is Dota 2 and Sid Meier's Civilization V.

### 8.4. Identifying these "All Star Players"

For these two games that you selected in 8.3 above, what is the user id of their respective "all star player"?

In [30]:
top_2_users = pd.merge(top_2, play, on=['game_name', 'hours_played_if_play'])[['game_name', 'user_id']]

In [31]:
top_2_users['user_id']

0     67694595
1    128065805
Name: user_id, dtype: int64

## 9. Learning about the All-Star Players

Finally, let's learn more about the play vs. purchase behavior of these two "all star" players that we identified in #8. Specifically for each of these two player user ids that you identified in #8, what percentage of games that they bought did they actually play?

In [32]:
AS_playPurchase = SS500_cleaned[SS500_cleaned['user_id'].isin(top_2_users['user_id'])]
game_count = AS_playPurchase.groupby(['user_id', 'activity'])['game_name'].nunique().unstack()

In [33]:
game_count['percent_played'] = (game_count['play']/game_count['purchase'])*100
game_count['percent_played']

user_id
67694595      63.157895
128065805    100.000000
Name: percent_played, dtype: float64